### PR TITLE
DOC Rework main encoding example

### LINKED ISSUE #341: Example 1: use DatetimeEncoder instead of isolating year
In [example 1](https://dirty-cat.github.io/stable/auto_examples/01_dirty_categories.html#sphx-glr-auto-examples-01-dirty-categories-py), we use the `year first hired` feature to predict wages. 

In the second part, when we use dirty_cat to encode variables, we may use the `DatetimeEncoder` instead of just passing through this variable. It is also possible not to drop the `date_first_hired`  variable and encode it instead.

To see what results this would give.

Also, would the SuperVectorizer (used in the second part) identify the `date_first_hired` column to be automatically encoded by the `DatetimeEncoder`? 
It does not at the moment for years.

### LINKED ISSUE #662: Outline for the main encoding example
(migrated from #657)

As discussed with Gaël, here is a proposed outline for the main encoding example. Please submit your suggestions!

- Start with a dirty dataset, probably employee_salaries
- First, use the TableVectorizer, just to say "with little trouble, it works ootb"
- Then go into what it did, and the internals (GapEncoder, DatetimeEncoder)
- Use the MinHashEncoder for a larger dataset (maybe the largest one we have)

Notes:
- Should we talk about the SimilarityEncoder?
- Should we introduce the problem of dirty categories in this notebook?
- We should integrate the [cross-validation example from](https://github.com/skrub-data/skrub/blob/3e312f9a41ff1e1fe4123e2af80ac93f15724b99/examples/encoding/grid_search_with_the_tablevectorizer.py) #546, but not sure where

cc @myungkim930

### PR DESCRIPTION
**Pull request by LilianBoulard**
Extracted from #546.

Fixes #341
Motivated by #662 (fixes #662)

### CODE REVIEW

File: examples/01_dirty_categories.py [hunk 1]
```diff
@@ -1,398 +1,154 @@
 """
```
Comment (GaelVaroquaux) [inline examples/01_dirty_categories.py:1]: We should change the file name

File: examples/01_dirty_categories.py [hunk 2]
```diff
@@ -1,398 +1,154 @@
...
+X.drop(labels=["underfilled_job_title"], axis="columns", inplace=True)
 
-###############################################################################
-# We still have a complex and heterogeneous dataframe:
 X
 
 ###############################################################################
-# The |TV| can to turn this dataframe into a form suited for
-# machine learning.
+# Let's extract a sample to have a more focused view: we'll limit our
```
Comment (GaelVaroquaux) [inline examples/01_dirty_categories.py:87]: I would like that we first run the analysis (pipelining TableVectorizer with HBGT before discussing this

Comment (GaelVaroquaux) [reply examples/01_dirty_categories.py:87]: Put it at the bottom, where we discuss the GapEncoder.

File: examples/01_dirty_categories.py [hunk 3]
```diff
@@ -1,398 +1,154 @@
...
-
-###############################################################################
-# Retrieving the feature importances:
 
-importances = regressor.feature_importances_
-std = np.std([tree.feature_importances_ for tree in regressor.estimators_], axis=0)
-indices = np.argsort(importances)
-# Sort from least to most
-indices = list(reversed(indices))
+# TODO
```
Comment (GaelVaroquaux) [inline examples/01_dirty_categories.py:144]: It seems to me that there is still a bit of work.

At least, we should put back the discussion on the feature names from the original example.

File: examples/01_dirty_categories.py [hunk 4]
```diff
@@ -1,398 +1,154 @@
...
-# --------------------------------------------
-#
-# In this section, we will train a regressor, and plot the feature importances.
+# We observe it has automatically assigned an appropriate encoder to
+# corresponding columns.
+# For example, it classified the columns 'gender', 'department',
+# 'department_name' and 'assignment_category' as low cardinality
+# string variables.
+# Two remarkable things:, it has affected a |GapEncoder| to the columns
+# `employee_position_title` and `division`, and a |DatetimeEncoder| to the
```
Comment (GaelVaroquaux) [inline examples/01_dirty_categories.py:134]: It seems that the replacement of the DatetimeEncoder did not work:
image

**Review by GaelVaroquaux — CHANGES_REQUESTED** [review]
A few suggestions of reformulation, and a couple of things to adress.

File: examples/01_encodings.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+plt.figure(figsize=(12, 9))
+plt.title("Feature importances")
+n = 20
+n_indices = indices[:n]
+labels = np.array(feature_names)[n_indices]
+plt.barh(range(n), importances[n_indices], color="b", yerr=std[n_indices])
+plt.yticks(range(n), labels, size=15)
+plt.tight_layout(pad=1)
+plt.show()
```
Comment (Vincent-Maladiere) [inline examples/01_encodings.py:177]: Nitpick, improves readability
```suggestion
top_indices = indices[:20]
labels = np.array(feature_names)[top_indices]

plt.figure(figsize=(12, 9))
plt.barh(
    y=labels,
    x=avg_importances[top_indices],
    yerr=std_importances[top_indices],
    color="b",
)
plt.yticks(fontsize=15)
plt.title("Feature importances")
plt.tight_layout(pad=1)
plt.show()
```

File: examples/01_encodings.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+regressor = RandomForestRegressor()
+
+pipeline = make_pipeline(TableVectorizer(), regressor)
+pipeline.fit(X, y)
+
+###############################################################################
+# We are retrieving the feature importances:
+
+importances = regressor.feature_importances_
+std = np.std([tree.feature_importances_ for tree in regressor.estimators_], axis=0)
```
Comment (Vincent-Maladiere) [inline examples/01_encodings.py:159]: Nitpick: improve readability
```suggestion
avg_importances = regressor.feature_importances_
std_importances = np.std([tree.feature_importances_ for tree in regressor.estimators_], axis=0)
```

**Review by Vincent-Maladiere — COMMENTED** [review]
Great concision effort, @LilianBoulard and @jovan-stojanovic (-200 lines!). This is way more user-friendly and straight to the point :)
Here is a review to extend this effort.

File: examples/01_encodings.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+X
+
+###############################################################################
+# We observe diverse columns in the dataset:
+#   - binary ('gender'),
+#   - numerical ('employee_annual_salary'),
+#   - categorical ('department', 'department_name', 'assignment_category'),
+#   - datetime ('date_first_hired')
+#   - dirty categorical ('employee_position_title', 'division').
```
[the remark this answers is not in the document]
Comment (jovan-stojanovic) [reply examples/01_encodings.py:62]: I followed what was recommended in #692, we might have a discussion later there to see what's best. Is that ok?

[the remark this answers is not in the document]
Comment (Vincent-Maladiere) [reply examples/01_encodings.py:62]: #692 suggests using e.g. ``'gender'``, written \``'gender'`` for datasets and column names, right?

[the remark this answers is not in the document]
Comment (jovan-stojanovic) [reply examples/01_encodings.py:62]: Ah ok, yes that seems to work. Will apply.

File: examples/01_encodings.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+pprint(tv.transformers_)
+
+###############################################################################
+# We observe it has automatically assigned an appropriate encoder to
+# corresponding columns:
+
+###############################################################################
+#     - The |OneHotEncoder| for low cardinality string variables, the columns
+#       'gender', 'department', 'department_name' and 'assignment_category'.
```
Comment (Vincent-Maladiere) [inline examples/01_encodings.py:95]:
```suggestion
#       ``'gender'``, ``'department'``, ``'department_name'`` and ``'assignment_category'``.
```

File: examples/01_encodings.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+tv.named_transformers_["low_card_cat"].get_feature_names_out()
+
+###############################################################################
+#     - The |GapEncoder| for high cardinality string columns, 'employee_position_title'
+#       and 'division'. The |GapEncoder| is a powerful encoder that can handle dirty
```
Comment (Vincent-Maladiere) [inline examples/01_encodings.py:101]:
```suggestion
#     - The |GapEncoder| for high cardinality string columns, ``'employee_position_title'``
#       and ``'division'``. The |GapEncoder| is a powerful encoder that can handle dirty
```

**Review by Vincent-Maladiere — COMMENTED** [review]
Last round of nitpick, and then we're good to go :)

**Review by Vincent-Maladiere — APPROVED** [review]
Thanks @jovan-stojanovic, LGTM! I think we can merge it as it is.

**Pull request merged**